src/backend/distributed/transaction/transaction_management.c

```diff
@@ -185,6 +192,63 @@ InCoordinatedTransaction(void)
 }
 
 
+/*
+ * Sets a flag to true indicating that the current node is executing a delegated
+ * function call, using forcePushdown, within a distributed transaction issued
+ * by the coordinator. Also, saves the distribution argument.
+ */
+void
+EnableInForceDelegatedFuncExecution(Const *distArgument)
+{
+	MemoryContext oldcontext = MemoryContextSwitchTo(TopTransactionContext);
+	elog(DEBUG1, "Saving Distribution Argument: %s",
+		 pretty_format_node_dump(nodeToString(distArgument)));
+	AllowedDistributionColumnValue = copyObject(distArgument);
+	ShouldAllowRestricted2PC = true;
+	MemoryContextSwitchTo(oldcontext);
+}
+
+
+void
+ResetAllowedShardKeyValue(void)
+{
+	AllowedDistributionColumnValue = NULL;
+}
+
+
+/*
+ * Returns true if a forcePushdown function distribution is executed in
+ * the current transaction.
+ */
+bool
+GetInForceDelegatedFuncExecution()
+{
+	return ShouldAllowRestricted2PC;
+}
+
+
+bool
+IsShardKeyValueAllowed(Const *shardKey)
+{
+	if (GetLocalGroupId() == COORDINATOR_GROUP_ID)
+	{
+		/* No check in the coordinator */
```

Comment:
With this PR, are we enabling function call delegation from workers? If so, then they could be delegated to the coordinator too.

Reply:
No, we shouldn't. We block that in **TryToDelegateFunctionCall**()

![image](https://user-images.githubusercontent.com/44680808/145624808-1dd2df57-4cc2-47c1-bc8f-e6efbd176f9d.png)

This function, which checks if we are using the same Shard key as the "distribution argument" of a delegated function. Such checks are not needed (or irrelevant) in the coordinator.

Reply:
This is the original code 

if (localGroupId != COORDINATOR_GROUP_ID || localGroupId == GROUP_ID_UPGRADING)

/* do not delegate from workers, or while upgrading */

Earlier, if the node is "worker" do not bother to delegate. In the new code, if the node is "worker" do not delegate only if the backend is Citus-initiated. But it has a minor bug, if the node is "coordinator", it will fail to return and proceed for delegation(again). I fixed it now